README.md
<p align="center">
<img width="300" alt="image" src="https://user-images.githubusercontent.com/35942268/226105723-86eb3042-fb51-4f51-b382-b833a952e9a2.png">
</p>

<p align="center">
   <a href="https://www.oscs1024.com/project/oscs/WGrape/lexer?ref=badge_small" alt="OSCS Status"><img src="https://www.oscs1024.com/platform/badge/WGrape/lexer.svg?size=small"/></a>
    <img src="https://img.shields.io/badge/JavaScript-ES5+-blue.svg">
    <img src="https://img.shields.io/npm/dt/chain-lexer.svg">
    <a href="https://app.travis-ci.com/github/WGrape/lexer"><img src="https://app.travis-ci.com/WGrape/lexer.svg?branch=main"><a>
    <img alt="GitHub release (latest by date)" src="https://img.shields.io/github/v/release/wgrape/lexer">
    <img src="https://img.shields.io/badge/Document-中文/English-orange.svg">
    <img src="https://img.shields.io/badge/License-MIT-green.svg">   
</p>

<div align="center">    
    <p>It is a lexical analyzer based on DFA that is built using JS and supports multi-language extensions. For a quick understanding and experience , please check the <a href="https://wgrape.github.io/lexer/">online website</a></p>
    <p>Document ：<a href="/README.zh-CN.md">中文</a> / <a href="/README.md">English</a></p>
</div>

<details>
  <summary>Contents</summary>

- [1、Background](#1)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(1) Situation](#11)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(2) Task](#12)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(3) Solution](#13)
- [2、Features](#2)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(1) Complete lexical analysis](#21)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(2) Support multi-language extension](#22)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(3) Provide state flow log](#23)
- [3、Get project](#3)
- [4、Ussage](#4)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(1) In your project](#41)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(2) Web preview and testing](#42)
- [5、Contributions](#5)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(1) Project Statistics](#51)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(2) Source code explanation](#52)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(3) Content contribution](#53)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(4) Release version](#54)
- &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[(5) Q&A](#55)
- [6、License](#6)

</details>

## <span id="1">1、Background</span>

### <span id="11">(1) Situation</span>

Most lexical analyzers are closely coupled with the language, the amount of code is relatively large. It's hard to pay attention to the essential principles of lexical analyzer.

### <span id="12">(2) Task</span>

In order to focus on the working principle of lexical analyzer , not to consider the small differences caused by different languages , an idea of making a ```lexer``` project that is completely decoupled from the language was born.

### <span id="13">(3) Solution</span>

```lexer``` through the following two files, realize the decoupling of lexical analyzer and language

- ```src/lexer.js``` is the core part of lexical analyzer within 300 lines, including ```ISR``` and ```DFA```
- ```src/lang/{lang}-define.js```is the language extension of lexical analyzer. Support different languages，such as ```src/lang/c-define.js```

## <span id="2">2、Features</span>

### <span id="21">(1) Complete lexical analysis</span>

From inputting the character sequence to generating ```token``` after the analysis, ```lexer``` has complete steps for lexical analysis, and 12 token types for most language extensions

<img width="850" alt="" src="https://user-images.githubusercontent.com/35942268/137583888-8c12a85c-4af7-4288-942f-d2a2fcfe30c6.png">

### <span id="22">(2) Support multi-language extension</span>

```lexer``` supports different language extensions such as ```Python```, ```Go```, etc. How to make different language extensions, please check [Contributions](#5)

- C ：A popular programming language，[click here](https://wgrape.github.io/lexer/?lang=c) to see its lexical analysis
- SQL ：A popular database query language，[click here](https://wgrape.github.io/lexer/?lang=sql) to see its lexical analysis
- Goal ：A goal parser problem from leetCode ，[click here](https://wgrape.github.io/lexer/?lang=goal) to see its lexical analysis

### <span id="23">(3) Provide state flow log</span>

The core mechanism of lexical analyzer is based on the state flow of ```DFA```. For this reason, ```lexer``` records detailed state flow log to achieve the following requirements of you

- Debug mode
- Automatically generate ```DFA``` state flow diagram

<img width="700" src="https://user-images.githubusercontent.com/35942268/136378451-e025fffd-425d-43f1-8a58-454a1011e9c3.png" />

## <span id="3">3、Get project</span>

After ```git clone``` command, no need for any dependencies, and no extra installation steps

## <span id="4">4、Ussage</span>

### <span id="41">(1) In your project</span>

If you need use ```lexer``` in your project, such as code editor, etc. 

#### Using NPM
```
npm install chain-lexer
```

```js
var chainLexer = require('chain-lexer');
let lexer = chainLexer.cLexer;

let stream = "int a = 10;";
lexer.start(stream);
let parsedTokens = lexer.DFA.result.tokens;

lexer = chainLexer.sqlLexer;
stream = "select * from test where id >= 10;";
lexer.start(stream);
parsedTokens = lexer.DFA.result.tokens;
```

#### Using Script
Import the ```package/{lang}-lexer.min.js``` file, then visit ```lexer``` variable to get the object of lexical analyzer，and visit ```lexer.DFA.result.tokens``` to get ```tokens```

```js
// 1. The code that needs lexical analysis
let stream = "int a = 10;";

// 2. Start lexical analysis
lexer.start(strem);

// 3. After the lexical analysis is done, get the generated tokens
let parsedTokens = lexer.DFA.result.tokens;

// 4. Do what you want to do
parsedTokens.forEach((token) => {
    // ... ...
});
```

The [Provide state flow log](#23) part in features，visit ```flowModel.result.paths``` will get the detail logs of state flow inside ```lexer```. The data format is as follows

```js
[
    {
        state: 0, // now state
        ch: "a", // read char
        nextSstate: 2, // next state
        match: true, // is match
        end: false, // is last char
    },
    // ... ...
]
```

### <span id="42">(2) Web preview and testing</span>

In order to preview the process of ```lexer``` in real time, to debug and test, there is a ```index.html``` file in the root directory of this project. Open it directly in your browser, and after entering the code will automatically output the ```Token``` generated after ```lexer``` analysis, as shown in the figure below

```c
int a = 10;
int b =20;
int c = 20;

float f = 928.2332;
char b = 'b';

if(a == b){
    printf("Hello, World!");
}else if(b!=c){
    printf("Hello, World! Hello, World!");
}else{
    printf("Hello!");
}
```

![img](https://user-images.githubusercontent.com/35942268/137584888-28a1ce09-3474-4158-8e6f-ccbdb8614930.gif)

or check the [online website](wgrape.github.io/lexer/)

## <span id="5">5、Contributions</span>

### <span id="51">(1) Project Statistics</span>

<a href="https://starchart.cc/WGrape/lexer"><img src="https://starchart.cc/WGrape/lexer.svg" width="700"></a>

### <span id="52">(2) Source code explanation</span>
Documents about source code development, project design, unit testing, automated testing, development specifications, and how to make extensions in different languages, please read [source code explanation](/doc/explain.md)

### <span id="53">(3) Content contribution</span>
- Add more new features
- Add more extensions ```/src/lang/{lang}-define.js```

### <span id="54">(4) Release version</span>
The project is released with the version number of ```A-B-C```，regarding release log, you can check the [CHANGELOG](./CHANGELOG.md) or the [release record](https://github.com/WGrape/lexer/releases)

- ```A```：Major upgrade
- ```B```：Minor upgrade
- ```C```：bug fix / features / ...

### <span id="55">(5) Q&A</span>
If you have any problems or questions, please [submit an issue](https://github.com/WGrape/lexer/issues/new)

## <span id="6">6、License</span>

![GitHub](https://img.shields.io/github/license/WGrape/lexer)


CHANGELOG.md
## [1.8.2](https://github.com/WGrape/lexer/compare/1.8.1...v1.8.2) (2022-09-19)

### Docs

* add directory fold ([581d3aa](https://github.com/WGrape/lexer/commit/581d3aa2c58f499a3bb5d84a813ea61fba85bd51))
* fix explain document ([c125698](https://github.com/WGrape/lexer/commit/c1256980a7756736d967ff07ec76a8e5b2f29d63))
* fix version ([5635a8a](https://github.com/WGrape/lexer/commit/5635a8ac3416d91a09ad3bcb4ae8b89b0bb9805d))

## [1.10.0](https://github.com/WGrape/lexer/compare/v1.9.0...v1.10.0) (2023-03-19)


### Features

* add compressLineFeed support ([3fb5970](https://github.com/WGrape/lexer/commit/3fb5970e192b738f4db857daf794072b91c0b412))


### Docs

* add npm operation ([049bb33](https://github.com/WGrape/lexer/commit/049bb33b2b9428e32e4852ec259856907c70506f))

## [1.9.0](https://github.com/WGrape/lexer/compare/v1.8.2...v1.9.0) (2023-03-18)


### Features

* add configure support and make some code cleaner ([bdde859](https://github.com/WGrape/lexer/commit/bdde8594961cabca869ad9e7603ec6a9b7bbc2bd))
* websit add comment feature ([5216151](https://github.com/WGrape/lexer/commit/521615151ae0235b9ef95b374d0e2ec26005abe6))


### Bug Fixes

* replace logo ([8ed5649](https://github.com/WGrape/lexer/commit/8ed5649b04ec37a4b1937f5f403f96858b665a4a))


### Styling

* adjuest websit comment style ([c57ef8b](https://github.com/WGrape/lexer/commit/c57ef8b8dd3d9d14ad086c214cecd2f25c66374c))


### Docs

* fix CHANGELOG.md ([9949b90](https://github.com/WGrape/lexer/commit/9949b9075a9bc9cd34a3d88a298cc9dfb99c518c))
* make doc better ([2d4f45c](https://github.com/WGrape/lexer/commit/2d4f45c17ba47cb6b8bb87b13875ab89d04384c8))
* make doc better ([9cb3bdd](https://github.com/WGrape/lexer/commit/9cb3bdd42541e1599922c5820a95db32101f3cc2))

### 6、Version ：1.8.1
Release date ：2021-12-20

desc ：Add string escape support

#### Features
- Feat: add string escape support, [here](https://github.com/WGrape/lexer/pull/47) is the PR

## [1.8.2](https://github.com/WGrape/lexer/compare/1.8.1...v1.8.2) (2022-09-19)


### Docs

* add directory fold ([581d3aa](https://github.com/WGrape/lexer/commit/581d3aa2c58f499a3bb5d84a813ea61fba85bd51))
* fix explain document ([c125698](https://github.com/WGrape/lexer/commit/c1256980a7756736d967ff07ec76a8e5b2f29d63))
* fix version ([5635a8a](https://github.com/WGrape/lexer/commit/5635a8ac3416d91a09ad3bcb4ae8b89b0bb9805d))

### 5、Version ：1.8.0
Release date ：2021-10-16

desc ：Added linefeed token and website support for real-time parsing

#### Features
- Feat: add a new token type to linefeed
- Feat: support real-time parsing on the [website](https://wgrape.github.io/lexer)

### 4、Version ：1.6.1
Release date ：2021-10-13

desc ：Change property name and remove ```goal_lexer``` property of chain-lexer, the package of NPM

#### Features
- Refactor: Change property name of chain-lexer.
  - ```c_lexer```: ```cLexer```
  - ```sql_lexer```: ```sqlLexer```
- Remove: remove ```goal_lexer``` property of chain-lexer

### 3、Version ：1.6.0
Release date ：2021-10-12

desc ：Publish project to NPM, named chain-lexer

#### Features
- NPM support: You can use lexer(chain-lexer) in your project by npm

### 2、Version ：1.5.0
Release date ：2021-10-10

desc ：Major upgrades to the project structure, such as package, shell, test, etc. You only need to import a ```/package/{lang}-lexer.min.js``` file in your project.

#### Features
- Add pack feature: pack ```/src/lexer.js``` and ```/src/lang/{lang}-define.js```
- Update testing: Decouple the testing from the ```/src/*``` file

### 1、Version ：1.0.0
Release date ：2021-09-31

desc ：the first version ```1.0.0``` of ```lexer``` is released.

#### Features
- Complete lexical analysis
- Support multi-language extension
- Provide state flow log


lang/sql-define.js
// Constant type: Enum
const ENUM_CONST = {
    // Simple Operator
    PLUS: '+',
    SUBTRACT: '-',
    TIMES: '*',
    DIVIDE: '/',
    MODE: '%',

    // Logical Operator
    NOT: '!',

    // Compare Operator
    GREAT: '>',
    LESS: '<',
    ASSIGN: '=',

    GREAT_EQUAL: ">=",
    LESS_EQUAL: "<=",
    EQUAL: "==",
    NOT_EQUAL: "!=",

    // Bit Operator
    BIT_AND: '&',
    BIT_OR: '|',
    BIT_NOT: '~',
    BIT_XOR: '^',

    CONDITION_AND: "&&",
    CONDITION_OR: "||",
    BIT_LEFT_SHIFT: "<<",
    BIT_RIGHT_SHIFT: ">>",

    // Whitespace mark
    WHITESPACE: ' ',

    // Bracket mark
    LEFT_BRACKET: '(',
    RIGHT_BRACKET: ')',
    LEFT_BRACE: '{',
    RIGHT_BRACE: '}',
    LEFT_SQUARE_BRACKET: '[',
    RIGHT_SQUARE_BRACKET: ']',

    // Other mark
    DOT: '.',
    COMMA: ',',
    SEMICOLON: ';',
    QUOTATION: '\'',
    DOUBLE_QUOTATION: '"',
    SHARP: '#',
    QUESTION: '?',
    COLON: ':',

    // TOKEN
    T_OPERATOR: 1,
    T_DOUBLE_OPERATOR: 2,
    T_SYMBOL: 3,
    T_WHITESPACE: 4,
    KEYWORD: 5,
    T_IDENTIFIER: 6,
    T_NUMBER: 7,
    T_FLOAT: 8,
    T_STRING: 9,
    T_CHAR: 10,
    T_UNKNOWN: 11,

    // DFA State
    S_RESET: 0,
    S_OPERATOR: 1,
    S_SYMBOL: 2,
    S_DOUBLE_CHAR_FIRST_BIT_AND: 3,
    S_DOUBLE_CHAR_SECOND_BIT_AND: 4,
    S_DOUBLE_CHAR_FIRST_BIT_OR: 5,
    S_DOUBLE_CHAR_SECOND_BIT_OR: 6,
    S_DOUBLE_CHAR_FIRST_GREAT: 7,
    S_DOUBLE_CHAR_SECOND_GREAT: 8,
    S_DOUBLE_CHAR_FIRST_LESS: 9,
    S_DOUBLE_CHAR_SECOND_LESS: 10,
    S_DOUBLE_CHAR_FIRST_NOT: 11,
    S_DOUBLE_CHAR_FIRST_ASSIGN: 12,
    S_DOUBLE_CHAR_SECOND_ASSIGN: 13,
    S_WHITESPACE: 14,
    S_IDENTIFIER: 15,
    S_NUMBER: 16,
    S_FLOAT: 17,
    S_STRING: 18,
    S_STRING_END: 19,
    S_CHAR: 20,
    S_CHAR_END: 21,
    S_END: 100,
};

// Constant type: Charset
const CHARSET_CONST = {
    // Char list
    CHAR: {
        // Operator list
        OPERATOR: [
            ENUM_CONST.PLUS, ENUM_CONST.SUBTRACT, ENUM_CONST.TIMES, ENUM_CONST.DIVIDE, ENUM_CONST.MODE,
            ENUM_CONST.NOT,
            ENUM_CONST.GREAT, ENUM_CONST.LESS, ENUM_CONST.ASSIGN,
            ENUM_CONST.BIT_AND, ENUM_CONST.BIT_OR, ENUM_CONST.BIT_NOT, ENUM_CONST.BIT_XOR,
        ],
        // Symbol list
        SYMBOL: [
            ENUM_CONST.LEFT_BRACKET, ENUM_CONST.RIGHT_BRACKET, ENUM_CONST.LEFT_BRACE, ENUM_CONST.RIGHT_BRACE, ENUM_CONST.LEFT_SQUARE_BRACKET, ENUM_CONST.RIGHT_SQUARE_BRACKET,
            ENUM_CONST.DOT, ENUM_CONST.COMMA, ENUM_CONST.SEMICOLON, ENUM_CONST.QUOTATION, ENUM_CONST.DOUBLE_QUOTATION, ENUM_CONST.SHARP, ENUM_CONST.QUESTION, ENUM_CONST.COLON,
        ],
        // Whitespace list
        WHITESPACE: [
            ENUM_CONST.WHITESPACE,
        ],
    },

    // Double char list
    DOUBLE_CHAR: [
        ENUM_CONST.GREAT_EQUAL, ENUM_CONST.LESS_EQUAL, ENUM_CONST.EQUAL, ENUM_CONST.NOT_EQUAL,
        ENUM_CONST.CONDITION_AND, ENUM_CONST.CONDITION_OR, ENUM_CONST.BIT_LEFT_SHIFT, ENUM_CONST.BIT_RIGHT_SHIFT,
    ],

    // Composition of double char: &&
    DOUBLE_CHAR_FIRST_BIT_AND: [ENUM_CONST.BIT_AND],
    DOUBLE_CHAR_SECOND_BIT_AND: [ENUM_CONST.BIT_AND],

    // Composition of double char: ||
    DOUBLE_CHAR_FIRST_BIT_OR: [ENUM_CONST.BIT_OR],
    DOUBLE_CHAR_SECOND_BIT_OR: [ENUM_CONST.BIT_OR],

    // Composition of double char: >>
    DOUBLE_CHAR_FIRST_GREAT: [ENUM_CONST.GREAT],
    DOUBLE_CHAR_SECOND_GREAT: [ENUM_CONST.GREAT],

    // Composition of double char: <<
    DOUBLE_CHAR_FIRST_LESS: [ENUM_CONST.LESS],
    DOUBLE_CHAR_SECOND_LESS: [ENUM_CONST.LESS],

    // Composition of double char: >= , <= , != , ==
    DOUBLE_CHAR_FIRST_NOT: [ENUM_CONST.NOT],
    DOUBLE_CHAR_FIRST_ASSIGN: [ENUM_CONST.ASSIGN],
    DOUBLE_CHAR_SECOND_ASSIGN: [ENUM_CONST.ASSIGN],

    // Keyword
    KEYWORD: [
        'select', 'from', 'where', 'and', 'group', 'by', 'order', 'desc', 'asc', 'limit',
        'add', 'constraint', 'alter', 'column', 'as', 'database', 'between', 'case', 'check',
        'create', 'alter', 'table', 'unique', 'index',
        'delete', 'distinct', 'drop', 'full', 'join', 'outer', 'inner', 'having', 'in', 'or', 'null',
        'insert', 'update', 'set', 'values', 'like', 'not', 'primary', 'into', 'top', 'union', 'all', 'view',
    ],
};

// Constant type: DFA State
const DFA_STATE_CONST = {
    // State of reset
    S_RESET: ENUM_CONST.S_RESET,

    // State of operator
    // State of symbol
    S_OPERATOR: ENUM_CONST.S_OPERATOR,
    S_SYMBOL: ENUM_CONST.S_SYMBOL,

    // State of double char: &&
    S_DOUBLE_CHAR_FIRST_BIT_AND: ENUM_CONST.S_DOUBLE_CHAR_FIRST_BIT_AND,
    S_DOUBLE_CHAR_SECOND_BIT_AND: ENUM_CONST.S_DOUBLE_CHAR_SECOND_BIT_AND,

    // State of double char: ||
    S_DOUBLE_CHAR_FIRST_BIT_OR: ENUM_CONST.S_DOUBLE_CHAR_FIRST_BIT_OR,
    S_DOUBLE_CHAR_SECOND_BIT_OR: ENUM_CONST.S_DOUBLE_CHAR_SECOND_BIT_OR,

    // State of double char: >>
    S_DOUBLE_CHAR_FIRST_GREAT: ENUM_CONST.S_DOUBLE_CHAR_FIRST_GREAT,
    S_DOUBLE_CHAR_SECOND_GREAT: ENUM_CONST.S_DOUBLE_CHAR_SECOND_GREAT,

    // State of double char: <<
    S_DOUBLE_CHAR_FIRST_LESS: ENUM_CONST.S_DOUBLE_CHAR_FIRST_LESS,
    S_DOUBLE_CHAR_SECOND_LESS: ENUM_CONST.S_DOUBLE_CHAR_SECOND_LESS,

    // State of double char: >=, <=, !=, ==
    S_DOUBLE_CHAR_FIRST_NOT: ENUM_CONST.S_DOUBLE_CHAR_FIRST_NOT,
    S_DOUBLE_CHAR_FIRST_ASSIGN: ENUM_CONST.S_DOUBLE_CHAR_FIRST_ASSIGN,
    S_DOUBLE_CHAR_SECOND_ASSIGN: ENUM_CONST.S_DOUBLE_CHAR_SECOND_ASSIGN,

    // State of whitespace
    S_WHITESPACE: ENUM_CONST.S_WHITESPACE,

    // State of identifier
    S_IDENTIFIER: ENUM_CONST.S_IDENTIFIER,

    // State of number literal
    S_NUMBER: ENUM_CONST.S_NUMBER,

    // State of float literal
    S_FLOAT: ENUM_CONST.S_FLOAT,

    // State of string literal
    S_STRING: ENUM_CONST.S_STRING,
    S_STRING_END: ENUM_CONST.S_STRING_END,

    // State of char literal
    S_CHAR: ENUM_CONST.S_CHAR,
    S_CHAR_END: ENUM_CONST.S_CHAR_END,

    // State of end
    S_END: ENUM_CONST.S_END,
};

// Tool package
let tool = {
    isUndefined(obj) {
        return typeof obj === "undefined";
    },
    isNodeEnvironment() {
        return typeof window === "undefined";
    },
    isInArray(needle, haystack) {
        return haystack.indexOf(needle) > -1;
    },
    judgeTokenTypeByValue(value) {
        if (tool.isInArray(value, CHARSET_CONST.CHAR.OPERATOR)) {
            return 'Operator';
        }
        if (tool.isInArray(value, CHARSET_CONST.DOUBLE_CHAR)) {
            return 'DoubleOperator';
        }
        if (tool.isInArray(value, CHARSET_CONST.CHAR.SYMBOL)) {
            return 'Symbol';
        }
        if (tool.isInArray(value, CHARSET_CONST.CHAR.WHITESPACE)) {
            return 'Whitespace';
        }
        if (tool.isInArray(value.toLowerCase(), CHARSET_CONST.KEYWORD)) {
            return 'Keyword';
        }
        return '';
    },
    judgeTokenType(state, value) {
        // Derive Token type according to Value
        let tokenType = tool.judgeTokenTypeByValue(value);
        if (tokenType) {
            return tokenType;
        }

        // Derive Token type according to State
        if (state === DFA_STATE_CONST.S_IDENTIFIER) {
            return 'Identifier';
        }
        if (state === DFA_STATE_CONST.S_NUMBER) {
            return 'Number';
        }
        if (state === DFA_STATE_CONST.S_FLOAT) {
            return 'Float';
        }
        if (state === DFA_STATE_CONST.S_STRING_END) {
            return 'String';
        }
        if (state === DFA_STATE_CONST.S_CHAR_END) {
            return 'String';
        }
        return 'Unknown';
    },
    getFirstCharState(ch) {
        if (tool.isInArray(ch, CHARSET_CONST.DOUBLE_CHAR_FIRST_BIT_AND)) {
            return DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_BIT_AND;
        }
        if (tool.isInArray(ch, CHARSET_CONST.DOUBLE_CHAR_FIRST_BIT_OR)) {
            return DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_BIT_OR;
        }
        if (tool.isInArray(ch, CHARSET_CONST.DOUBLE_CHAR_FIRST_GREAT)) {
            return DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_GREAT;
        }
        if (tool.isInArray(ch, CHARSET_CONST.DOUBLE_CHAR_FIRST_LESS)) {
            return DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_LESS;
        }
        if (tool.isInArray(ch, CHARSET_CONST.DOUBLE_CHAR_FIRST_NOT)) {
            return DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_NOT;
        }
        if (tool.isInArray(ch, CHARSET_CONST.DOUBLE_CHAR_FIRST_ASSIGN)) {
            return DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_ASSIGN;
        }
        return DFA_STATE_CONST.S_RESET;
    },
    getSecondCharState(ch) {
        if (tool.isInArray(ch, CHARSET_CONST.DOUBLE_CHAR_SECOND_BIT_AND)) {
            return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_BIT_AND;
        }
        if (tool.isInArray(ch, CHARSET_CONST.DOUBLE_CHAR_SECOND_BIT_OR)) {
            return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_BIT_OR;
        }
        if (tool.isInArray(ch, CHARSET_CONST.DOUBLE_CHAR_SECOND_GREAT)) {
            return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_GREAT;
        }
        if (tool.isInArray(ch, CHARSET_CONST.DOUBLE_CHAR_SECOND_LESS)) {
            return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_LESS;
        }
        if (tool.isInArray(ch, CHARSET_CONST.DOUBLE_CHAR_SECOND_ASSIGN)) {
            return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_ASSIGN;
        }
        return DFA_STATE_CONST.S_RESET;
    },

    isOperatorChar(ch) {
        return tool.isInArray(ch, CHARSET_CONST.CHAR.OPERATOR);
    },
    isSymbolChar(ch) {
        return tool.isInArray(ch, CHARSET_CONST.CHAR.SYMBOL);
    },
    isWhitespaceChar(ch) {
        return tool.isInArray(ch, CHARSET_CONST.CHAR.WHITESPACE);
    },
    isFirstCharOfDoubleChar(ch) {
        return tool.getFirstCharState(ch) > 0;
    },
    isSecondCharOfDoubleChar(ch) {
        return tool.getSecondCharState(ch) > 0;
    },
    isAlphabetChar(ch) {
        return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z');
    },
    isNumberChar(ch) {
        return !isNaN(parseInt(ch)) && ch >= 0 && ch <= 9;
    },
    isInStates(state, states) {
        return tool.isInArray(state, states);
    }
};

// Define the DFA state flow model
let flowModel = {
    result: {
        paths: [],
    },
    resultChange: {
        pathGrow(path) {
            flowModel.result.paths.push(path)
        },
        toDefault() {
            flowModel.result.paths = [];
        }
    },

    getNextState(ch, state, matchs) {

        // Handling of common parts
        if (tool.isInStates(state, [DFA_STATE_CONST.S_STRING])) {
            if (ch !== ENUM_CONST.DOUBLE_QUOTATION) {
                return DFA_STATE_CONST.S_STRING;
            } else {
                return DFA_STATE_CONST.S_STRING_END;
            }
        }
        if (tool.isInStates(state, [DFA_STATE_CONST.S_CHAR])) {
            if (ch === ENUM_CONST.QUOTATION) {
                return DFA_STATE_CONST.S_CHAR_END;
            }
            return DFA_STATE_CONST.S_CHAR;
        }

        // Handling of non-common parts
        if (tool.isAlphabetChar(ch)) {
            if (tool.isInStates(state, [DFA_STATE_CONST.S_RESET, DFA_STATE_CONST.S_IDENTIFIER])) {
                return DFA_STATE_CONST.S_IDENTIFIER;
            }
        } else if (tool.isNumberChar(ch)) {
            if (tool.isInStates(state, [DFA_STATE_CONST.S_RESET, DFA_STATE_CONST.S_NUMBER])) {
                return DFA_STATE_CONST.S_NUMBER;
            }
            if (tool.isInStates(state, [DFA_STATE_CONST.S_FLOAT])) {
                return DFA_STATE_CONST.S_FLOAT;
            }
            if (tool.isInStates(state, [DFA_STATE_CONST.S_IDENTIFIER])) {
                return DFA_STATE_CONST.S_IDENTIFIER;
            }
        } else if (tool.isOperatorChar(ch)) {
            if (tool.isFirstCharOfDoubleChar(ch)) {
                if (tool.isInStates(state, [DFA_STATE_CONST.S_RESET])) {
                    return tool.getFirstCharState(ch);
                }
            }
            if (tool.isSecondCharOfDoubleChar(ch)) {
                if (state === DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_BIT_AND && ch === ENUM_CONST.BIT_AND) {
                    return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_BIT_AND;
                }
                if (state === DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_BIT_OR && ch === ENUM_CONST.BIT_OR) {
                    return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_BIT_OR;
                }
                if (state === DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_GREAT && ch === ENUM_CONST.GREAT) {
                    return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_GREAT;
                }
                if (state === DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_LESS && ch === ENUM_CONST.LESS) {
                    return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_LESS;
                }

                if (state === DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_GREAT && ch === ENUM_CONST.ASSIGN) {
                    return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_ASSIGN;
                }
                if (state === DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_LESS && ch === ENUM_CONST.ASSIGN) {
                    return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_ASSIGN;
                }
                if (state === DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_NOT && ch === ENUM_CONST.ASSIGN) {
                    return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_ASSIGN;
                }
                if (state === DFA_STATE_CONST.S_DOUBLE_CHAR_FIRST_ASSIGN && ch === ENUM_CONST.ASSIGN) {
                    return DFA_STATE_CONST.S_DOUBLE_CHAR_SECOND_ASSIGN;
                }
            }

            if (tool.isInStates(state, [DFA_STATE_CONST.S_RESET])) {
                return DFA_STATE_CONST.S_OPERATOR;
            }
        } else if (tool.isSymbolChar(ch)) {
            if (tool.isInStates(state, [DFA_STATE_CONST.S_RESET])) {
                if (ch === ENUM_CONST.QUOTATION) {
                    return DFA_STATE_CONST.S_CHAR;
                }
                if (ch === ENUM_CONST.DOUBLE_QUOTATION) {
                    return DFA_STATE_CONST.S_STRING;
                }
                return DFA_STATE_CONST.S_SYMBOL;
            }
            if (tool.isInStates(state, [DFA_STATE_CONST.S_NUMBER]) && ch === ENUM_CONST.DOT) {
                return DFA_STATE_CONST.S_FLOAT;
            }
        } else if (tool.isWhitespaceChar(ch)) {
            if (tool.isInStates(state, [DFA_STATE_CONST.S_RESET])) {
                return DFA_STATE_CONST.S_WHITESPACE;
            }
        } else {
            if (tool.isInStates(state, [DFA_STATE_CONST.S_RESET, DFA_STATE_CONST.S_IDENTIFIER])) {
                return DFA_STATE_CONST.S_IDENTIFIER;
            }
        }
        return DFA_STATE_CONST.S_RESET;
    },
};

// Unit Testing
if (tool.isNodeEnvironment()) {
    let assert = require('assert');
    assert.equal(tool.isUndefined(flowModel.FakeValue), true, "tool.isUndefined单测失败");
}

// Automated Testing
let autoTest = {
    returnCaseList() {
        return [
            {
                "input": "",
                "output": 0,
            },
            {
                "input": "select",
                "output": 1,
            },
            {
                "input": "select*",
                "output": 2,
            },
            {
                "input": "create table;",
                "output": 3,
            },
            {
                "input": ";;select*;select",
                "output": 6,
            },
            {
                "input": "select*from table where name ='hello';",
                "output": 9,
            },
            {
                "input": "select id,name,student.age from student where name = \"张三\" and score=23.22; ",
                "output": 19,
            },
            {
                "input": "select*from test where(id)>=2 and id<=10 ( or name = 'hello');",
                "output": 21,
            },
            {
                "input": "select* from t where (id)>2 or id != 3 and age between 10 and 20 order by id,name asc limit 2,10;",
                "output": 31,
            },
            {
                "input": "select id,name,student.age from student where (id<22) and (id>2) group by name having id > 20 order by id desc limit 0,10;",
                "output": 38,
            },
            {
                "input": "select * from table where id >= 2 and id != 2 or id == 2;",
                "output": 17,
            },
        ];
    },
};